In [19]:
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

In [21]:
n = 1000

x = np.linspace(0, 2*np.pi , n)

x1 = x**3
x2 = np.sin(x)


mu= 1
sigma= np.sqrt(0.2)
noise = np.random.normal(mu, sigma ,n)

y = 2 - x1 + 3*x2 + noise


rows = []





In [22]:

X_a = x.reshape(-1, 1)**2
X_train, X_test, y_train, y_test = train_test_split(X_a, y, test_size=0.3, random_state=42)

model_a_raw = LinearRegression(fit_intercept=False)
model_a_raw.fit(X_train, y_train - 2)

y_pred_train = model_a_raw.predict(X_train) + 2
y_pred_test = model_a_raw.predict(X_test) + 2

mse_train = mean_squared_error(y_train, y_pred_train)
rmse_train = np.sqrt(mse_train)
ndei_train = rmse_train / np.std(y_train)

mse_test = mean_squared_error(y_test, y_pred_test)
rmse_test = np.sqrt(mse_test)

ndei_test = rmse_test / np.std(y_test)

rows.append({
    "Model": "(a) x^2, b=2 fixed",
    "Scaling": "Raw",
    "w1": float(model_a_raw.coef_[0]),
    "b": 2.0,
    "MSE_train": mse_train, "RMSE_train": rmse_train, "NDEI_train": ndei_train,
    "MSE_test": mse_test, "RMSE_test": rmse_test, "NDEI_test": ndei_test
})



In [23]:
scaler = StandardScaler()
X_train_s = scaler.fit_transform(X_train)
X_test_s = scaler.transform(X_test)

model_a_scaled = LinearRegression(fit_intercept=False)
model_a_scaled.fit(X_train_s, y_train - 2)

y_pred_train = model_a_scaled.predict(X_train_s) + 2
y_pred_test = model_a_scaled.predict(X_test_s) + 2

mse_train = mean_squared_error(y_train, y_pred_train)
rmse_train = np.sqrt(mse_train)
ndei_train = rmse_train / np.std(y_train)

mse_test = mean_squared_error(y_test, y_pred_test)
rmse_test = np.sqrt(mse_test)
ndei_test = rmse_test / np.std(y_test)

rows.append({
    "Model": "(a) x^2, b=2 fixed",
    "Scaling": "Scaled",
    "w1": float(model_a_scaled.coef_[0]),
    "b": 2.0,
    "MSE_train": mse_train, "RMSE_train": rmse_train, "NDEI_train": ndei_train,
    "MSE_test": mse_test, "RMSE_test": rmse_test, "NDEI_test": ndei_test
})



In [24]:
X_b = np.column_stack((x1, x2))
X_train, X_test, y_train, y_test = train_test_split(X_b, y, test_size=0.3, random_state=42)

model_b = LinearRegression(fit_intercept=False)
model_b.fit(X_train, y_train)

y_pred_train = model_b.predict(X_train)
y_pred_test = model_b.predict(X_test)

mse_train = mean_squared_error(y_train, y_pred_train)
rmse_train = np.sqrt(mse_train)
ndei_train = rmse_train / np.std(y_train)

mse_test = mean_squared_error(y_test, y_pred_test)
rmse_test = np.sqrt(mse_test)
ndei_test = rmse_test / np.std(y_test)

rows.append({
    "Model": "(b) x1,x2 no bias",
    "Scaling": "Raw",
    "w1": float(model_b.coef_[0]),
    "w2": float(model_b.coef_[1]),
    "b": 0.0,
    "MSE_train": mse_train, "RMSE_train": rmse_train, "NDEI_train": ndei_train,
    "MSE_test": mse_test, "RMSE_test": rmse_test, "NDEI_test": ndei_test
})



In [25]:
X_c = X_b
X_train, X_test, y_train, y_test = train_test_split(X_c, y, test_size=0.3, random_state=42)

model_c = LinearRegression(fit_intercept=True)
model_c.fit(X_train, y_train)

y_pred_train = model_c.predict(X_train)
y_pred_test = model_c.predict(X_test)

mse_train = mean_squared_error(y_train, y_pred_train)
rmse_train = np.sqrt(mse_train)
ndei_train = rmse_train / np.std(y_train)

mse_test = mean_squared_error(y_test, y_pred_test)
rmse_test = np.sqrt(mse_test)
ndei_test = rmse_test / np.std(y_test)

rows.append({
    "Model": "(c) x1,x2 + bias",
    "Scaling": "Raw",
    "w1": float(model_c.coef_[0]),
    "w2": float(model_c.coef_[1]),
    "b": float(model_c.intercept_),
    "MSE_train": mse_train, "RMSE_train": rmse_train, "NDEI_train": ndei_train,
    "MSE_test": mse_test, "RMSE_test": rmse_test, "NDEI_test": ndei_test
})





In [26]:
mu_values = [0.0, 1.0, 2.0]
sigma2_values = [0.05, 0.2, 1.0]

for mu_d in mu_values:
    for sigma2_d in sigma2_values:
        sigma_d = np.sqrt(sigma2_d)
        noise_d = np.random.normal(mu_d, sigma_d, n)
        y_d = 2 - x1 + 3*x2 + noise_d

        X_d = np.column_stack((x1, x2))
        X_train, X_test, y_train, y_test = train_test_split(X_d, y_d, test_size=0.3, random_state=42)

        model_d = LinearRegression(fit_intercept=True)
        model_d.fit(X_train, y_train)

        y_pred_train = model_d.predict(X_train)
        y_pred_test = model_d.predict(X_test)

        mse_train = mean_squared_error(y_train, y_pred_train)
        rmse_train = np.sqrt(mse_train)
        ndei_train = rmse_train / np.std(y_train)

        mse_test = mean_squared_error(y_test, y_pred_test)
        rmse_test = np.sqrt(mse_test)
        ndei_test = rmse_test / np.std(y_test)

        rows.append({
            "Model": "(d) noise change on (c)",
            "mu": mu_d,
            "sigma2": sigma2_d,
            "Scaling": "Raw",
            "w1": float(model_d.coef_[0]),
            "w2": float(model_d.coef_[1]),
            "b": float(model_d.intercept_),
            "MSE_train": mse_train, "RMSE_train": rmse_train, "NDEI_train": ndei_train,
            "MSE_test": mse_test, "RMSE_test": rmse_test, "NDEI_test": ndei_test
        })



In [27]:
df_results = pd.DataFrame(rows)

cols_order = [c for c in ["Model", "Scaling", "mu", "sigma2", "w1", "w2", "b",
                          "MSE_train", "RMSE_train", "NDEI_train",
                          "MSE_test", "RMSE_test", "NDEI_test"] if c in df_results.columns]
df_results = df_results[cols_order]

df_results.round(4)

Model Scaling   mu  sigma2       w1      w2       b  \
0        (a) x^2, b=2 fixed     Raw  NaN     NaN  -5.2310     NaN  2.0000   
1        (a) x^2, b=2 fixed  Scaled  NaN     NaN -70.0503     NaN  2.0000   
2         (b) x1,x2 no bias     Raw  NaN     NaN  -0.9710  4.8859  0.0000   
3          (c) x1,x2 + bias     Raw  NaN     NaN  -0.9998  2.9882  2.9876   
4   (d) noise change on (c)     Raw  0.0    0.05  -1.0001  2.9957  2.0084   
5   (d) noise change on (c)     Raw  0.0    0.20  -1.0002  2.9900  2.0121   
6   (d) noise change on (c)     Raw  0.0    1.00  -1.0004  2.8957  2.0092   
7   (d) noise change on (c)     Raw  1.0    0.05  -1.0002  2.9879  3.0108   
8   (d) noise change on (c)     Raw  1.0    0.20  -1.0002  2.9615  3.0275   
9   (d) noise change on (c)     Raw  1.0    1.00  -1.0006  2.8851  3.0416   
10  (d) noise change on (c)     Raw  2.0    0.05  -1.0002  2.9877  4.0220   
11  (d) noise change on (c)     Raw  2.0    0.20  -1.0000  2.9649  3.9883   
12  (d) noise change on (c)     Raw  2.0    1.00  -1.0012  2.9928  4.0739   

    MSE_train  RMSE_train  NDEI_train   MSE_test  RMSE_test  NDEI_test  
0    262.8194     16.2117      0.2287   278.4641    16.6872     0.2253  
1   3805.6816     61.6902      0.8703  3886.4232    62.3412     0.8417  
2      3.8693      1.9671      0.0278     4.0299     2.0075     0.0271  
3      0.2077      0.4557      0.0064     0.2032     0.4508     0.0061  
4      0.0481      0.2192      0.0031     0.0524     0.2290     0.0031  
5      0.2092      0.4574      0.0065     0.1768     0.4205     0.0057  
6      1.0787      1.0386      0.0147     1.1228     1.0596     0.0143  
7      0.0454      0.2131      0.0030     0.0483     0.2199     0.0030  
8      0.1986      0.4457      0.0063     0.1833     0.4281     0.0058  
9      1.1270      1.0616      0.0150     1.0513     1.0253     0.0139  
10     0.0446      0.2112      0.0030     0.0560     0.2367     0.0032  
11     0.1865      0.4319      0.0061     0.1947     0.4412     0.0060  
12     1.0174      1.0087      0.0142     1.0320     1.0159     0.0137

MSE: خطای مربعی

RMSE: خطا با واحد y

پس قابل تفسیر انسانی می‌شه:

«مدل من به طور میانگین حدوداً ۵ واحد اشتباه می‌کند.»
چرا RMSE خوب است اینجا؟

تو نویز با واریانس‌های مختلف گذاشتی: ۰.۰۵، ۰.۲، ۱

RMSE مستقیماً نشان می‌دهد که خطای عددی چقدر زیاد شده؛
یعنی «حس واقعی» می‌دهد که خطا تحت تغییر نویز چقدر شده.

می‌بینی که هم‌زمان با افزایش نویز، «اشتراک عددی خطا» زیاد می‌شود.


NDEI: خطای نرمال‌شده

I ≈ نسبتِ «خطای مدل» به «پراکندگی طبیعی داده»

اگر NDEI ≈ ۱ → یعنی خطای مدل اندازه‌ی پراکندگی خود داده است
→ مدل چیزی بهتر از «حدسِ متوسط y» نیست.

اگر NDEI ≪ ۱ (مثلاً ۰.۱) → مدل دقت خیلی خوبی دارد.

عدد بدون واحد است → فقط نسبت است.


این‌جا اگر فقط RMSE ببینی، نمی‌فهمی:

«آیا مدل نسبت به نویز بدتر شده، یا فقط خود سیگنال پرنوسان‌تر شده؟»

NDEI کمک می‌کند:

اگر RMSE ↑ و std(y) هم ↑ و NDEI تقریباً ثابت → مدل هنوز «به نسبت سیگنال» خوب است.

اگر NDEI هم ↑ → یعنی مدل نسبت به نویز عملکرد بدتری دارد.


MSE معیار پایه‌ای مبتنی بر مربع خطاست و برای بهینه‌سازی تحلیلی (OLS) مناسب است، ولی به مقیاس y وابسته است.

RMSE با گرفتن ریشه‌ی دوم MSE، خطا را در همان واحد خروجی مدل ارائه می‌دهد و از نظر تفسیر فیزیکی/مهندسی قابل فهم‌تر است، اما همچنان به مقیاس y وابسته است.

NDEI با نرمال‌سازی RMSE بر حسب انحراف معیار خروجی، یک معیار بدون‌واحد ارائه می‌کند که امکان مقایسه‌ی مدل‌ها در شرایط مختلف (مثلاً نویزهای متفاوت یا تغییر مقیاس y) را فراهم می‌کند. مقدار NDEI نزدیک ۱ نشان‌دهنده‌ی عملکردی مشابه پیش‌بینی میانگین است، درحالی‌که مقادیر بسیار کوچک‌تر از ۱ بیانگر مدل دقیق و مناسب هستند.



مدل (ج):

NDEI_test ≈ ۰.۰۰۶ → فوق‌العاده کوچیک یعنی تقریباً perfectly fitted (چون ساختار مدل با ساختار داده منطبق است).

مدل (الف) Scaled:

NDEI_test ≈ ۰.۸–۰.۹ → یعنی مدل ضعیف، خطا تقریباً اندازه‌ی تغییرات طبیعی داده است.







الف)

وقتی فقط یک فیچر دارد
و بایاس ثابته

نرمال‌سازی نه‌تنها کمک نمی‌کنه، بلکه ضرر هم می‌زنه
که دقیقاً تو خروجی‌ات هم دیده می‌شه (MSE بدتر شده)


نرمال‌سازی (StandardScaler) زمانی مفید است که:

چند فیچر داریم

مقیاس فیچرها خیلی با هم فرق دارد

مدل باید بین آن‌ها تعادل عددی برقرار کند

با نرمال‌سازی:

مدل باید:

هم اثر حذف x₁ را جبران کند

هم اثر تغییر مقیاس مصنوعی x₂

هم بدون اجازه تغییر بایاس این کار را بکند

نتیجه؟

وزن w بیش‌ازحد بزرگ می‌شود
خطا (MSE, RMSE, NDEI) بالا می‌رود


ب)

y=w1x1+w2x2

نتیجه:
w1 ≈ -0.97
w2 ≈ 4.88

چون مدل مجبور شده اثر بایاس 2 رو با ضرایب جبران کنه

خطا نسبت به مدل (ج) بالاتره

ولی نسبت به مدل (الف) خیلی بهتره


ج)

مدل:

𝑦=𝑤1𝑥1+𝑤2𝑥2+𝑏
y=w1x1+w2x2​+b

نتایج عالی:
w1 ≈ -1
w2 ≈ 3
b  ≈ 2+(μ=1) =3

چون نویز میانگینش 1 بوده



د)
تحلیل نتایج:
 تغییر μ:

فقط b تغییر می‌کند

w1 و w2 تقریباً ثابت می‌مانند

 افزایش σ²:

MSE ↑

RMSE ↑

NDEI ↑

اما بایاس مدل خراب نمی‌شود

 این دقیقاً رفتار مورد انتظار یک مدل OLS سالم است